### PR TITLE
System Identification toolkit for Ardupilot compatible vehicles: GSoC 2024

### PR DESCRIPTION
**Pull request by Astik-2002**
This pr presents a web based toolkit for transfer function and statespace model identification for Ardupilot compatible UAVs. UAV flight logs are analysed between given timestamps to calculate frequency response, from which the mathematical models are identified. Final bug fixes and user review remains for this webtool, hence some commits may follow

### CODE REVIEW

File: SysID/index.html [hunk 1]
```diff
@@ -0,0 +1,949 @@
...
+      <input type="text" id="customDenominator" name="customDenominator"><br><br>
+
+      <label for="tf_params">Symbolic params:</label>
+      <input type="text" id="tf_params" name="tf_params"><br><br>
+
+      <div id="plotDiv"></div>
+
+    </div>
+
+    <!-- FOR STATE SPACE MODEL -->
```
Comment (bnsgeyer) [inline SysID/index.html:72]: Add capability for user to specify start and stop frequencies
Add capability to apply LPF to the input and all outputs

File: SysID/index.html [hunk 2]
```diff
@@ -0,0 +1,949 @@
...
+            bnd = tuple(bounds_array)
+            con_str = list(con_str)
+
+            g = 9.81
+
+
+            resampled_input = time_seq_preprocess(time_seq_source, input_data[0])
+            resampled_output = time_seq_preprocess(time_seq_source, output_data[0])
+
+            simo_iden = FreqIdenSIMO(time_seq_source,2, 10, input_data[0], output_data[0], win_num=None)
```
Comment (bnsgeyer) [inline SysID/index.html:880]: Use the start and stop frequencies as input into this call.

**Review by bnsgeyer — COMMENTED** [review]
@Astik-2002 I added my comments here about what we discussed and also additional ones that I found.  This will help us track the issues.
As for the x8SSM_yaw.py and x8yaw.csv files, I would just delete them as they can't be used in this webtool environment.

**Review by IamPete1 — COMMENTED** [review]
Needs a bit of formatting love. Split out the two inline python functions to there own files and put everything except the setup javascript into its a dedicated js file.

File: SysID/index.html [hunk 3]
```diff
@@ -0,0 +1,949 @@
...
+
+      <label for="number of params">Number of params:</label>                  
+      <input type="number" id="num_params" name="Inputs">
+
+      <label for="number of constraints">Number of Constraints:</label>                  
+      <input type="number" id="num_cons" name="Inputs">
+      
+      
+      <label for="start time">Start time (s):</label>
+      <input type="number" id="starttime" name="starttime"><br><br>
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply SysID/index.html]: fixed

File: SysID/index.html [hunk 4]
```diff
@@ -0,0 +1,949 @@
...
+
+      <label for="end time">End time (s):</label>
+      <input type="number" id="endtime" name="endtime"><br><br>
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply SysID/index.html]: Fixed

File: SysID/index.html [hunk 5]
```diff
@@ -0,0 +1,949 @@
...
+              b, a = butter(order, normal_cutoff, btype='low', analog=False)
+              return b, a
+
+
+            def apply_lowpass_filter(data, cutoff, fs, order=5):
+              b, a = butter_lowpass(cutoff, fs, order=order)
+              y = filtfilt(b, a, data)
+              return y
+
+            time_seq_source = np.array(time_data).flatten()/1000000
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply SysID/index.html]: Fixed

File: SysID/index.html [hunk 6]
```diff
@@ -0,0 +1,949 @@
...
+      <label for="start time">Start time (s):</label>
+      <input type="number" id="starttime" name="starttime"><br><br>
+
+      <label for="end time">End time (s):</label>
+      <input type="number" id="endtime" name="endtime"><br><br>
+
+      <button id="createFieldsButton" type="button">Submit</button>
+      <br><br>
+
+      <div id="inputFieldsContainer"></div>
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply SysID/index.html:97]: Only applied to outputs.  Need to add to inputs as well

File: SysID/index.html [hunk 7]
```diff
@@ -0,0 +1,949 @@
...
+          console.log("order A: ", orderA);
+
+          if (isNaN(numOutputs) || numOutputs <= 0) {
+            alert('Please enter valid numbers for inputs and outputs.');
+            return;
+          }
+          const inputValues = getFieldValues('input', 1);
+          const outputValues = getFieldValues('output', numOutputs);
+          const symbolic_var = getParamValues(numParams);
+          const t_start = parseFloat(document.getElementById('starttime').value.trim(), 10);
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply SysID/index.html]: seems to be fixed with just the variable name change

File: SysID/index.html [hunk 8]
```diff
@@ -0,0 +1,949 @@
...
+          const t_end = parseFloat(document.getElementById('endtime').value.trim(), 10);
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply SysID/index.html]: Seems to be fixed with just the variable name change

File: SysID/index.html [hunk 9]
```diff
@@ -0,0 +1,949 @@
...
+
+            F = getMatrixJs(matrixA, orderA, orderA, sym_var)
+            G = getMatrixJs(matrixB, orderA, numInputs, sym_var)
+            H0 = getMatrixJs(matrixH0, numOutputs, orderA, sym_var)
+            H1 = getMatrixJs(matrixH1, numOutputs, orderA, sym_var)
+            
+            t_start = float(t_start)
+            t_end = float(t_end)
+
+            time_seq_source = np.array(time_data).flatten()/1000000
```
[the remark this answers is not in the document]
Comment (bnsgeyer) [reply SysID/index.html]: fixed

**Comment by IamPete1** [comment]
Merged as part of https://github.com/ArduPilot/WebTools/pull/181. Thanks for all your work this summer!

**Pull request closed without merging**